### ISSUE TITLE
Using TypedSelect in elementselect

### ISSUE THREAD

**Issue by enyachoke** [viritin/viritin#62]
First thanks for promptly answering my previous question. Now I have another issue.
Am trying to use ElementCollectionField with a jpa onetomany relationship.
Here are my classes.
EstimationManual.java
```java
@Entity
public class EstimationManual extends AbstractEntity {
    private String description;

    public String getDescription() {
        return description;
    }

    public void setDescription(String description) {
        this.description = description;
    }
    @OneToMany(cascade = CascadeType.ALL, mappedBy = "estimationManual",fetch = FetchType.EAGER)
    private List<EstimationManualItem> estimationManualItems = new ArrayList<>();

    public List<EstimationManualItem> getEstimationManualItems() {
        return estimationManualItems;
    }

    public void setEstimationManualItems(List<EstimationManualItem> estimationManualItems) {
        this.estimationManualItems = estimationManualItems;
    }


}
```
EstimationManualItem.java
```java
@Entity
public class EstimationManualItem extends AbstractEntity {
    public EstimationManual getEstimationManual() {
        return estimationManual;
    }

    public void setEstimationManual(EstimationManual estimationManual) {
        this.estimationManual = estimationManual;
    }

    @ManyToOne
    private EstimationManual estimationManual;
    @ManyToOne
    private Resource resource;
    @Min(0)
    @NotNull
    private Double quantity = 1.0;

    public Double getQuantity() {
        return quantity;
    }

    public void setQuantity(Double quantity) {
        this.quantity = quantity;
    }

    public Resource getResource() {
        return resource;
    }

    public void setResource(Resource resource) {
        this.resource = resource;
    }
    public EstimationManualItem() {
    }
}
```
EstimationManualForm.java
```java
@org.springframework.stereotype.Component
public class EstimationManualForm extends AbstractForm<EstimationManual> {
    TextField description = new MTextField("Description");
    public static class EstimationManualItemRow {
        TypedSelect resource = new TypedSelect<>().withWidth("6em");
        TextField quantity = new MTextField().withInputPrompt("Quantity");
    }
    ElementCollectionField<EstimationManualItem> estimationManualItems = new ElementCollectionField<>(
            EstimationManualItem.class, EstimationManualItemRow.class).withCaption("Estimation Manual Items");
    @Override
    protected Component createContent() {
        return new MVerticalLayout(
                new MHorizontalLayout(
                new MFormLayout(
       				//Add form fields
                        description
                ).withMargin(false)
                ),
                estimationManualItems,
                getToolbar()
        );
    }
    public EstimationManualForm() {
        setEagerValidation(true);
    }
    public EstimationManualForm(EstimationManual estimationManual) {
        setEagerValidation(true);
    }
}
```
When using MultiSelectTable I can use setOption method to load items to select. How can I achieve the same for this usecase.
Maybe I should document this for people struggling with the same.

**Comment by mstahv** [comment]
Hi,

Do you have existing referred objects ( of type EstimationManualItem ) that you'd like to remap to the edited EstimationManual, or would you like to create new ones?

If you just want to allow adding new items you can override createInstance method or pass in Instantiator interface to wire the back reference to the correct "master entity". Then on persist/save things should be saved just fine to DB. But in this kind of case with modern JPA 2 you should check if ElementCollection/Embedded would work for you.

For picking existing entities there is now ready solution. I'd probably crete a "choose existing" button with a popup Window listing possible options with Table/Grid. Then on item click you could add the selected entity to be added to ElementCollectionField for editing.

cheers,
matti

**Comment by enyachoke** [comment]
ElementCollection won't work I need to select already existing  entities. Via a onetomany relationship. So I should probably try the second solution. Jst curious can I use vaadin components in a viritin form.

**Comment by mstahv** [comment]
Yes, of course, I do that a lot, e.g. with DataField's. With basic TextField the "eager validation" don't work though.

**Comment by enyachoke** [comment]
Can't try it out yet but can something link this.
```java 
cb_resource = new ComboBox("Resource");

for(resource : ListResource)
{
    cb_resource.addItem(resource.getId);
    cb_resource.setItemcaption(resource.getId,resource.getName);
}
```

**Comment by mstahv** [comment]
So you'd use that to add existing items to ElementCollectionField? That could work indeed, but I'd actually use TypedSelect from Viritin there, better typing and thats why an easier API. There is also LazyComboBox that will save a lot of cpu, memory and slow queries in case your EstimationManualItem table grows to tens of thousands of items.

**Comment by enyachoke** [comment]
Okay maybe am dumb but after several hours I can't really get this on working. Let me describe the problem a bit. One of the fields in the EstimationManualItem element collection (resource) has to selected from an existing inventory to populate the EstimationManualItem resource (ManytoOne ). I could really appreciate an example on how you could achieve. My project has many places it require this kind of flow.

**Comment by mstahv** [comment]
Hi, don't feel dump, these are not easy things, and we wave way too little real world examples. I like to help you a you seem to have basics in very good shape so the "hard parts" you face are definitely something we should improve some how. 

Do you want to allow editing those referred entities in the UI (when editing the master entity)? If not, I'd just use a table based solution, more specifically MultiSelectTable from Viritin. The best solution depends on more detailed characteristics, such as is there a really lot of those referred items or just some? If there is a huge set of available options, table based approach is not that good. You'd need a "selected list" and a separate search view to pick those referred entities.

If you project is open source, would you share the source code? It might be easier that way to suggest something more specific.

**Comment by enyachoke** [comment]
Thanks here is the code in involved
I have already shared the EstimationManual.java with is the parent class and EstimationManualItem.java which is used in the element collection. The only missing model is Resource which is what needs to be selected to fill the reource field of the EstimationManualItem model.
Resource.java
```java
@Entity
@Table(name = "resource")
public class Resource extends AbstractEntity {

    @Column(name="resourceType")
    @NotNull(message = "This field can't be null")
    private ResourceType resourceType;

    @Column(name="resourceName")
    @NotNull(message = "This field can't be null")
    private String resourceName;

    @Column(name="description")
    private String description;

    @OneToMany(cascade = CascadeType.ALL, mappedBy = "estimationManual",fetch = FetchType.EAGER)
    private List<EstimationManualItem> estimationManualItems = new ArrayList<>();


    public List<EstimationManualItem> getEstimationManualItems() {
        return estimationManualItems;
    }

    public void setEstimationManualItems(List<EstimationManualItem> estimationManualItems) {
        this.estimationManualItems = estimationManualItems;
    }



    public ResourceType getResourceType() {
        return resourceType;
    }

    public void setResourceType(ResourceType resourceType) {
        this.resourceType = resourceType;
    }


    public String getResourceName() {
        return resourceName;
    }

    public void setResourceName(String resourceName) {
        this.resourceName = resourceName;
    }

    public String getDescription() {
        return description;
    }

    public void setDescription(String description) {
        this.description = description;
    }


    public enum  ResourceType{
        Labour,Material,Equipment
    }
```
Repo
EstimationManualRepository.java
```java
public interface EstimationManualRepository extends JpaRepository<EstimationManual,Long> {
}
```
Currently the system looks like
![Image of Yaktocat](https://s3.eu-central-1.amazonaws.com/harambesa/010cc21d-7d50-4fc1-bac4-786d2cb5ace9/new+folder/Screenshot+from+2015-03-06+11%3A46%3A36.png)

You can see the drop down is empty and that  is what I need to select. The system is a construction erp so the resources are varios equipment,labour and materials which can hardly be more than 500. Hope this is enough to work with.

**Comment by mstahv** [comment]
Hi,

Now that you posted the screenshot I actually realised I was looking at the wrong relation and probably gave some really weird comments above :-) 

The ElementCollectionField actually had a concept that would allow some configuration and overridding the way how "editors" are created, but that was pretty well hidden. The snapshot version now has API to actually set that. I created a separate branch to my spring data vaadin crud with pretty similar entity structure and a usage example:

https://github.com/mstahv/spring-data-vaadin-crud/blob/multiple-entities/src/main/java/crud/vaadin/PhoneBookEntryForm.java#L42-L51

I hope this helps!

cheers,
matti
